server/src/main/java/org/elasticsearch/search/fetch/FetchPhase.java

```diff
@@ -120,25 +120,29 @@ private SearchHits buildSearchHits(SearchContext context, Profiler profiler) {
         List<FetchSubPhaseProcessor> processors = getProcessors(context.shardTarget(), fetchContext, profiler);
         NestedDocuments nestedDocuments = context.getSearchExecutionContext().getNestedDocuments();
 
-        int currentReaderIndex = -1;
-        LeafReaderContext currentReaderContext = null;
+        List<LeafReaderContext> leaves = context.searcher().getIndexReader().leaves();
         LeafNestedDocuments leafNestedDocuments = null;
         CheckedBiConsumer<Integer, FieldsVisitor, IOException> fieldReader = null;
         boolean hasSequentialDocs = hasSequentialDocs(docs);
         SourceLoader.Leaf leafSourceLoader = null;
+        int leafIndex = -1;
+        LeafReaderContext leafReaderContext = null;
+        int endReaderIdx = -1;
         for (int index = 0; index < context.docIdsToLoadSize(); index++) {
             if (context.isCancelled()) {
                 throw new TaskCancelledException("cancelled");
             }
             int docId = docs[index].docId;
             try {
-                int readerIndex = ReaderUtil.subIndex(docId, context.searcher().getIndexReader().leaves());
-                if (currentReaderIndex != readerIndex) {
+                if (index >= endReaderIdx) {
                     profiler.startNextReader();
                     try {
-                        currentReaderContext = context.searcher().getIndexReader().leaves().get(readerIndex);
-                        currentReaderIndex = readerIndex;
-                        if (currentReaderContext.reader()instanceof SequentialStoredFieldsLeafReader lf
+                        leafIndex++;
+                        leafIndex = ReaderUtil.subIndex(docId, leaves.subList(leafIndex, leaves.size())) + leafIndex;
+                        leafReaderContext = context.searcher().getIndexReader().leaves().get(leafIndex);
+                        endReaderIdx = endReaderIdx(context, leafReaderContext, index, docs);
+                        int[] docIdsInLeaf = docIdsInLeaf(index, endReaderIdx, docs, leafReaderContext.docBase);
+                        if (leafReaderContext.reader()instanceof SequentialStoredFieldsLeafReader lf
```

Comment:
nit: spacing before `instanceof`

Reply:
You can't! There's a bug in our formatter that needs, like, a fix upstream. The delivery folks are managing it. But, for now, you *can't* have a space there!